### PR TITLE
CH17255: Create s3_delete_first transport type

### PR DESCRIPTION
**Pull request by fcraposo**
Create a new S3 transport type that deletes all the files in the folder before initializing the consumer.

### CODE REVIEW

**Review by velppa — CHANGES_REQUESTED** [review]
Few suggestions:
1. Don't need to have new transport type, just extend the constructor with an optional parameter
2. Version change should be in at least minor number, that's not a bugfix change

File: analytics/client.py
Comment (inigofh) [inline analytics/client.py:63]: I agree with @pavel-popov, let's move the `delete_first` as a parameter of the `_init_` constructor and set it to `False` by default for `s3`.

**Review by inigofh — APPROVED** [review]
Looks good to me. I've just left a comment similar to @pavel-popov's suggestion.

**Comment by fcraposo** [comment]
> Few suggestions:
> 
> 1. Don't need to have new transport type, just extend the constructor with an optional parameter

That was my first choice, but then I would have to change the signature of the Client constructor by adding a param that would only be related to the S3Consumer, not the standard Consumer, and that could be misleading. 

This is why I preferred to use a new transport type, instead. 

Do you still prefer to move it to the Client constructor, @pavel-popov and @inigofh ?

> 2. Version change should be in at least minor number, that's not a bugfix change

Well spotted. I'll change.

File: analytics/client.py
Comment (fcraposo) [reply analytics/client.py:63]: That was my first choice, but then I would have to change the signature of the Client constructor by adding a param that would only be related to the S3Consumer, not the standard Consumer, and that could be misleading.

This is why I preferred to use a new transport type, instead.

Do you still prefer to move it to the Client constructor?

Comment (inigofh) [reply analytics/client.py:63]: I'm fine with the way you did it. It could be nicer/simpler to have it the other way, but I don't think it's a blocker. For me the PR is ok.

**Comment by velppa** [comment]
Creating a new transport in order to set additional options doesn't seem right.

> but then I would have to change the signature of the Client constructor by adding a param that would only be related to the S3Consumer

And that's expected, we should go even further and set suffix and s3 bucket right from the client, and remove all those endpoints from `analytics-api` - https://github.com/FindHotel/analytics-api/blob/f7e07e5211b64f3a5ad685721e657c7ffae31ecf/analytics.yaml.j2#L239-L282

**Comment by fcraposo** [comment]
> Creating a new transport in order to set additional options doesn't seem right.

It's a compromise not to clutter the Client constructor with Consumer specific parameters. And, in practice, it's just a slug that abstracts a different usage of the same `S3Consumer` class. 

Ideally, we could have some sort of ConsumerFactory that would return the correct instance for the type of transport. But that would be a larger and breaking change. 

My goal is to implement the `delete_first` strategy in a quick, non-breaking way, without cluttering/hacking the Client constructor even more...

**Comment by fcraposo** [comment]
> > but then I would have to change the signature of the Client constructor by adding a param that would only be related to the S3Consumer
> 
> And that's expected, we should go even further and set suffix and s3 bucket right from the client, and remove all those endpoints from `analytics-api` - https://github.com/FindHotel/analytics-api/blob/f7e07e5211b64f3a5ad685721e657c7ffae31ecf/analytics.yaml.j2#L239-L282

Not sure if I understood your suggestion, but I think that it's maybe outside the scope of what we need to achieve right now.

**Comment by fcraposo** [comment]
I'll deploy it as is and we can discuss and improve the solution.

**Comment by germangh** [comment]
@fcraposo merging a PR when there are request for changes by any of the reviewers should be done only in very exceptional circumstances. And when the request for changes comes from the repo maintainer then you should just not do it. I will document this shortly so that it's clear to everyone what is the github etiquette that we should all follow and what is the meaning of a "maintainer".

In the case of this repo @pavel-popov is the maintainer (you can see it in the repo description in github). The reason why he is the maintainer is because he wrote all the findhotel-specific code here.  You can ask him to add you to the list of maintainers and then it will be his call whether he wants to give you that power or not. If you want you can always fork the repo and maintain your own version of it if for whatever reason Pavel's repo doesn't fit your needs and you feel blocked. Not recommended in general, but there can be reasons to do this sometimes. It's @inigofh and your call which code you use in the systems you own. 

Since you already merged this PR and also created a new release in Pypi, the preferable option is that you:

* Comply with the requests from Pavel then create a new release, update your code to use the new release and then optionally you can decide to delete the deprecated release from Pypi since right now there is a very high probability that you are the only user of that release so it's an opportunity to clean up. @pavel-popov I think it'd be best if the release to Pypi happens automatically during the CI. Note: @fcraposo I am going to remove you from the list of maintainers in Pypi since the maintainers in Pypi should be the same as the maintainers of this repo. 

The 2nd option is that you decide to fork the SDK and take ownership of that fork and then you follow your own standards and whatever approach you prefer to develop the SDK. But this should only be done when there are very good reasons for it, such your team being continuously blocked by the repo maintainer not attending to PRs you make and that you need to achieve your team's objectives.

**Comment by fcraposo** [comment]
Ok. 

Sorry if I seemed dismissive, @pavel-popov and @germangh. That was not my intention. I only thought that it wouldn't be a problem to deploy this minor change to unblock me on my current task and postpone the ideal solution a little bit. 

About your suggestion, @pavel-popov,

I was under the impression that we needed the endpoints to authorize via the `x-api-key`. You're suggesting that we specify the full path in the Client and the `s3` transport would simply upload it directly?

**Comment by velppa** [comment]
Adding a new transport is not a minor change and having a new transport for configuring some client options doesn't seem right - as I mentioned before.

I’ve put a bit of thought on proposed changes and conclude it to be a bad idea – to give EvAS _client_ ability to delete files from the EvAS _system_ – from guarantees (immutability) and security angles (for example, how to protect from human errors to unintentionally delete files).

The access configuration for EvAS S3 bucket [here](https://github.com/FindHotel/analytics-eventsources/blob/d5db60ee8fc6902dbd5bb99d7dfa99b167b1c5e4/deployment/batch.tf#L35-L43) doesn't look right - application shouldn't define write/delete access to resources for itself - it the owner system responsibility (EvAS) to manage access to its resources - [like here](https://github.com/FindHotel/analytics-api/blob/f7e07e5211b64f3a5ad685721e657c7ffae31ecf/deployment/main.tf#L109-L120).

For this use-case - downloading CA reports - the application needs to operate on some state (files on S3) - read/write/delete - we don’t need EvAS SDK for that, better use S3 SDK directly and hold the files to be managed by the application (not by EvAS).

I think that using EvAS for downloading reports from CA platform was the wrong design decision and we should change it, for example, this comment from @inigofh is very valid - https://github.com/FindHotel/analytics-eventsources/pull/102#issuecomment-643660363 and we need to delete data from Snowflake instead.

What I would do:

0. Delete `1.7.0` release from PyPI and either delete this PR from Github or make this repo to be private
1. Rename `analytics-eventsources` to a better name, like `ca-reporting`
2. Have a separate s3 bucket where the app keeps its data (and manages it accordingly), like `fh-ca-reporting`
3. Delete `package` event stream from EvAS and all associated endpoints, delete `report` pipe and `delivery_prod.report` table from Snowflake.
4. Rewrite code that reads from `delivery_prod.report` to read from Snowflake external stages that point to `fh-ca-reporting` bucket

**Pull request merged**
